### PR TITLE
test: Code Coverage: Add Edit Expense #9

### PR DESCRIPTION
**Pull request by jayfyle**
### <samp>🤖 Generated by Copilot at 28c00e0</samp>

This pull request refactors the code and tests for the add-edit-expense page and adds mock data for various features. It improves the type safety, code quality, readability, and maintainability of the `add-edit-expense.page.ts` file and the `add-edit-expense-5.spec.ts` file. It also adds mock data for custom input fields, default expense values, tax settings, tax groups, and report options in the `custom-input.data.ts`, `default-txn-field-values.data.ts`, `org-settings.data.ts`, `tax-group.data.ts`, and `report-unflattened.data.ts` files.

### <samp>🤖 Generated by Copilot at 28c00e0</samp>

> _Sing, O Muse, of the skillful coder who refactored_
> _The mock data files and the add-edit-expense page_
> _To test the custom input fields and the tax calculations_
> _And to enhance the quality and readability of his code._

### <samp>🤖 Generated by Copilot at 28c00e0</samp>

*  Refactor mock data files and add new mock data for custom input, tax settings, and report options features ([link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-ffa8f022a8cce95188ae7c79b4e1f34074d6c8086297a6822f542110666d2405R1-R56), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-88d03c9481ee9f872719d0aa13033ce30c0e50a9d5de1403d54b362a3f5282cbR7-R14), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-1cac010b51c1a7c2fbdf9bf224a7e267779365abe08b0f16df1ef46e41483951R1207-R1235), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-d67cba8ad3a3712af6f510fa26bd7076e3e8bcd472aef15dcd35f0a4cdc32eb2R1558-R1804), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-7017b79f4cdec2ca4579722329b79907c450d94744ee97e4f39ded9798d0fa8bR23-R49))
* Update import statements in `add-edit-expense-5.spec.ts` to use the refactored mock data ([link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639L10-R22), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639L29-R43))
* Add new test cases for `getCustomFields` and `ionViewWillEnter` methods in `add-edit-expense-5.spec.ts` to test the custom input and initialization features and increase code coverage ([link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639L81-R95), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-ba304f41b9ceb84ec3edd25c39133e53d238dc01ad0cb7b04fbd999aa2300639R827-R1020))
* Fix type error in `individualProjectIds$` variable in `add-edit-expense.page.ts` ([link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL195-R195))
* Refactor `ionViewWillEnter` method in `add-edit-expense.page.ts` by moving some code to new methods `initSubjectObservables`, `setupSelectedProjectObservable`, `setupSelectedCostCenterObservable`, `getCCCpaymentMode`, and `clearCategoryOnValueChange` to improve readability and maintainability ([link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2391-R2391), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2404-R2447), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2447-R2492), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2469-R2508), [link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2769-R2806))
* Refactor `getCustomFields` method in `add-edit-expense.page.ts` by adding new methods `getProjectDependentFields` and `getCostCenterDependentFields` to improve readability and maintainability ([link](https://github.com/fylein/fyle-mobile-app/pull/2201/files?diff=unified&w=0#diff-492aac4c505a9ef8fb102b40f692554c3d0dc493b7abddc97b6665166efcc83bL2969-R3000))

## Clickup
https://app.clickup.com/t/85zt9ph25

## Code Coverage
`87.23% Statements 1258/1442`
`72.66% Branches 1119/1540`
`87.06% Functions 377/433`
`87.52% Lines 1221/1395`

## UI Preview
Please add screenshots for UI changes

### CODE REVIEW

File: src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts [hunk 1]
```diff
@@ -812,5 +824,199 @@ export function TestCases5(getTestBed) {
...
+
+      component.getCustomFields().subscribe((res) => {
+        expect(res).toEqual(customInputData1);
+        expect(customFieldsService.standardizeCustomFields).toHaveBeenCalledOnceWith([], dependentCustomFields);
+        expect(component.getProjectDependentFields).toHaveBeenCalledTimes(1);
+        expect(component.getCostCenterDependentFields).toHaveBeenCalledTimes(1);
+      });
+    });
+
+    describe('ionViewWillEnter():', () => {
```
Comment (jayfyle) [inline src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts:872]: I will add the remaining tests in the next PR

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 2]
```diff
@@ -2388,7 +2388,7 @@ export class AddEditExpensePage implements OnInit {
     }
   }
 
-  ionViewWillEnter() {
+  initSubjectObservables() {
```
Comment (Julias0) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:2391]: Why call these subject observables?

File: src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts [hunk 3]
```diff
@@ -812,5 +824,251 @@ export function TestCases5(getTestBed) {
...
+        project_dependent_fields: [],
+      });
+
+      component.fg.controls.project_dependent_fields.setValue(projectDependentFields);
+
+      const result = component.getProjectDependentFields();
+      expect(result).toEqual(projectDependentFields);
+    });
+
+    it('getCostCenterDependentFields(): should get project dependent fields', () => {
```
Comment (Dimple16) [inline src/app/fyle/add-edit-expense/add-edit-expense-5.spec.ts:839]: `should get cost center  dependent fields`

**Review by Dimple16 — CHANGES_REQUESTED** [review]
added comments

File: src/app/core/mock-data/custom-input.data.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
+import { CustomInput } from '../models/custom-input.model';
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/mock-data/custom-input.data.ts:1]: No that is a service related data file from the previous guidelines. I will have to  segregate the data, the file does not 
seem to have any data relevant to the model

File: src/app/core/mock-data/report-unflattened.data.ts [hunk 5]
```diff
@@ -1555,3 +1555,250 @@ export const reportOptionsData2 = [
     },
   },
 ];
+
+export const reportOptionsData3 = [
```
[the remark this answers is not in the document]
Comment (jayfyle) [reply src/app/core/mock-data/report-unflattened.data.ts:1559]: There was no model for this piece of data. I have created a new model and shifted these to a new file.

**Review by Julias0 — CHANGES_REQUESTED** [review]
Will review after Dimple

File: src/app/core/models/v1/default-txn-field-values.model.ts [hunk 6]
```diff
@@ -1,8 +1,10 @@
+import { ExpenseField } from './expense-field.model';
+
 export interface DefaultTxnFieldValues {
   bus_travel_class: string;
   distance_unit: string;
-  flight_journey_travel_class: string;
-  flight_return_travel_class: string;
+  flight_journey_travel_class: string | Partial<ExpenseField>;
```
Comment (Dimple16) [inline src/app/core/models/v1/default-txn-field-values.model.ts:6]: are we sure about this? is the API returning expense field object?

**Review by Dimple16 — APPROVED** [review]
one question, rest lgtm

File: src/app/core/models/v1/default-txn-field-values.model.ts (hunk 6 above)
Comment (jayfyle) [reply src/app/core/models/v1/default-txn-field-values.model.ts]: I have reverted the change, it will be a string.

**Review by Julias0 — APPROVED** [review]

**Pull request merged**